full-moon/src/tokenizer.rs

```diff
@@ -650,260 +642,148 @@ impl<'a> fmt::Display for StringLiteralQuoteType {
-    ))(code)
-}
-
-fn parse_number(code: &str) -> IResult<&str, &str> {
-    alt((
-        #[cfg(feature = "roblox")]
-        parse_roblox_number,
-        parse_hex_number,
-        parse_basic_number,
-        parse_no_int_fractional_number,
-    ))(code)
-}
+type RawToken<'a> = Result<TokenType<'a>, TokenizerErrorType>;
 
-fn advance_number(code: &str) -> Advancement {
-    match parse_number(code) {
-        Ok((_, number)) => Ok(Some(TokenAdvancement {
-            advance: number.chars().count(),
-            token_type: TokenType::Number {
-                text: Cow::from(number),
-            },
-        })),
-        Err(_) => Ok(None),
+impl<'a> From<TokenType<'a>> for RawToken<'a> {
+    fn from(token_type: TokenType<'a>) -> RawToken<'a> {
+        Ok(token_type)
     }
 }
 
-#[inline]
-fn parse_identifier(code: &str) -> IResult<&str, &str> {
-    recognize(pair(
-        alt((alpha1, tag("_"))),
-        many0(alt((alphanumeric1, tag("_")))),
-    ))(code)
-}
-
-fn advance_identifier(code: &str) -> Advancement {
-    match parse_identifier(code) {
-        Ok((_, identifier)) => Ok(Some(TokenAdvancement {
-            advance: identifier.chars().count(),
-            token_type: TokenType::Identifier {
-                identifier: Cow::from(identifier),
-            },
-        })),
-        Err(_) => Ok(None),
+impl<'a> From<TokenizerErrorType> for RawToken<'a> {
+    fn from(error: TokenizerErrorType) -> RawToken<'a> {
+        Err(error)
     }
 }
 
-#[inline]
-fn parse_shebang(code: &str) -> IResult<&str, &str> {
-    recognize(pair(tag("#!"), many_till(anychar, line_ending)))(code)
-}
-
-fn advance_shebang(code: &str) -> Advancement {
-    match parse_shebang(code) {
-        Ok((_, line)) => Ok(Some(TokenAdvancement {
-            advance: line.chars().count(),
-            token_type: TokenType::Shebang {
-                line: Cow::from(line),
-            },
-        })),
-        Err(_) => Ok(None),
-    }
-}
-
-#[inline]
-fn parse_multi_line_string_start(code: &str) -> IResult<&str, &str> {
-    delimited(tag("["), take_while(|x: char| x == '='), tag("["))(code)
-}
-
-#[inline]
-fn parse_single_line_string(code: &str) -> IResult<&str, Option<(StringLiteralQuoteType, &str)>> {
-    macro_rules! quoted {
-        ($quote:literal, $quote_type:path) => {
-            pair(
-                success($quote_type),
-                delimited(
-                    tag($quote),
-                    alt((
-                        escaped(none_of(concat!("\r\n\\", $quote)), '\\', anychar),
-                        success(""),
-                    )),
-                    tag($quote),
-                ),
+peg::parser! {
+    grammar tokens() for str {
+        use super::ParseSymbol;
+        use peg::ParseLiteral;
+        use super::StringLiteralQuoteType as QuoteType;
+        rule line_ending()
+            = "\n" / "\r\n"
+        rule space()
+            = [' '|'\t']
+        pub(super)
```

Comment:
I know what pub(super) does, but what does it do in this context?

And for these, can you change it to `pub(super) rule whitespace() -> RawToken<'input>` so it's obvious what it's attached to?

Reply:
> I know what pub(super) does, but what does it do in this context?

`peg::parser!` creates a module, so `pub(super)` exposes the rule to the containing (i.e. `tokenizer`) module as a function.